x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/codec/KMeansClustering.java

```diff
@@ -0,0 +1,299 @@
+            } else if (random.nextDouble() < pcentroidsCount * (1.0 / numDocs)) {
+                int c = random.nextInt(pcentroidsCount);
+                float[][] pvectors = decodeVector(bytes, pqCount, pdims);
+                for (int pq = 0; pq < pqCount; pq++) {
+                    pcentroids[pq][c] = pvectors[pq];
+                }
+            }
+            numDocs++;
+        }
+
+        System.out.println("Running k-means with on [" + numDocs + "] docs...");
+        if (algorithm == 1) { // Lloyds algorithm
+            for (int pq = 0; pq < pqCount; pq++) {
+                docCentroids[pq] = bigArrays.newByteArray(numDocs);
+            }
+            for (int itr = 0; itr < numIters; itr++) {
+                float fraction = (itr < numIters - 1) && numDocs > 100_000 ? sampleFraction : 1.0f;
+                pcentroids = runKMeansStepLloyds(itr, fraction, valuesProducer, field, pcentroids, docCentroids);
+            }
+        } else { // KMeans sort algorithm
+            double[][][] cdists = new double[pqCount][pcentroidsCount][pcentroidsCount]; // inter-centroid distances
+            int[][][] cdistIndexes = new int[pqCount][pcentroidsCount][pcentroidsCount]; // indexes of other centroids sorted by distances
+            for (int pq = 0; pq < pqCount; pq++) {
+                docCentroids[pq] = bigArrays.newByteArray(numDocs);
+                docCentroidDists[pq] = bigArrays.newDoubleArray(numDocs);
+            }
+            for (int itr = 0; itr < numIters; itr++) {
+                float fraction = (itr < numIters - 1) && numDocs > 100_000 ? sampleFraction : 1.0f;
+                pcentroids = runKMeansStepSort(itr, fraction, valuesProducer, field,
+                    pcentroids, docCentroids, docCentroidDists, cdists, cdistIndexes) ;
+            }
+        }
+    }
+
+    /**
+     * Runs one iteration of k-means Lloyds. For each document vector, we first find the
+     * nearest centroid, then update the location of the new centroid.
+     */
+    private static float[][][] runKMeansStepLloyds(int iter,
+            float sampleFraction,
+            DocValuesProducer valuesProducer,
+            FieldInfo field,
+            float[][][] centroids,
+            ByteArray[] docCentroids) throws IOException {
+
+        int pqCount = centroids.length; // number of quantizers
+        int cCount = centroids[0].length; // number of product centroids in each quantizer
+        int pdims = centroids[0][0].length; // number of dims in each product centroid
+        float[][][] newCentroids = new float[pqCount][cCount][pdims];
+        int[][] newCentroidSize = new int[pqCount][cCount];
+
+        double distToCentroid = 0.0;
+        double distToOtherCentroids = 0.0;
+        int doc;
+        int numDocs = 0;
+        BinaryDocValues values = valuesProducer.getBinary(field);
+        while ((doc = values.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+            if (random.nextFloat() > sampleFraction) {
```

Comment:
We should compute the sample only once, otherwise we're learning each step on a different dataset ?